src/tail/session.rs

```diff
@@ -36,41 +48,96 @@ impl Session {
                 let tail_id = success.result.id;
 
                 // Loop indefinitely to send "heartbeat" to API and keep log streaming alive.
+                // This should loop forever until SIGINT is issued or Wrangler process is killed
+                // through other means.
+                let duration = Duration::from_millis(1000 * KEEP_ALIVE_INTERVAL);
+                let mut delay = delay_for(duration);
+
                 loop {
-                    thread::sleep(Duration::from_secs(60));
-                    let heartbeat_result = send_heartbeat(target, &client, &tail_id).await;
-                    if heartbeat_result.is_err() {
-                        return heartbeat_result;
+                    match rx.try_recv() {
+                        Err(TryRecvError::Empty) => {
+                            if delay.is_elapsed() {
+                                let heartbeat_result =
+                                    send_heartbeat(&target, &client, &tail_id).await;
+                                if heartbeat_result.is_err() {
+                                    return heartbeat_result;
+                                }
+                                delay = delay_for(duration);
+                            }
+                        }
+                        _ => {
+                            return Ok(());
+                        }
```

Comment:
i'm interested in this - the flow isn't super obvious to me, why is the heartbeat sent on this error and why do we stop if nothing is received?

Reply:
i'll add a comment